File: LLMClient/Platforms/Android/MlcLlm/MlcLlmEngine.java
package com.llmclient.mlcllm;

import android.content.Context;
import android.util.Log;

import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;

import java.io.File;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Consumer;

/**
 * MLC LLM Engine wrapper for Android
 * Provides high-performance GPU-accelerated LLM inference using OpenCL
 */
public class MlcLlmEngine {
    private static final String TAG = "MlcLlmEngine";

    // Native library handle
    private long engineHandle = 0;
    private boolean isLoaded = false;
    private String modelPath = "";
    private ExecutorService executor;

    // Callbacks
    public interface StreamCallback {
        void onToken(String token);
        void onComplete(String fullResponse);
        void onError(String error);
    }

    public interface ProgressCallback {
        void onProgress(double progress, String status);
    }

    static {
        try {
            // Load MLC LLM native libraries
            System.loadLibrary("mlc_llm");
            System.loadLibrary("tvm_runtime");
            Log.i(TAG, "MLC LLM native libraries loaded successfully");
        } catch (UnsatisfiedLinkError e) {
            Log.e(TAG, "Failed to load MLC LLM libraries: " + e.getMessage());
        }
    }

    public MlcLlmEngine() {
        executor = Executors.newSingleThreadExecutor();
    }

    /**
     * Initialize the engine with a model
     * @param context Android context
     * @param modelId Model identifier (e.g., "Qwen2-1.5B-Instruct-q4f16_1-MLC")
     * @param modelPath Path to model files
     */
    public void initialize(Context context, String modelId, String modelPath) {
        this.modelPath = modelPath;
        executor.execute(() -> {
            try {
                engineHandle = nativeCreateEngine(modelPath, getDefaultConfig());
                if (engineHandle != 0) {
                    isLoaded = true;
                    Log.i(TAG, "MLC LLM Engine initialized with model: " + modelId);
                } else {
                    Log.e(TAG, "Failed to create MLC LLM engine");
                }
            } catch (Exception e) {
                Log.e(TAG, "Error initializing engine: " + e.getMessage());
            }
        });
    }

    /**
     * Check if model is loaded and ready
     */
    public boolean isReady() {
        return isLoaded && engineHandle != 0;
    }

    /**
     * Generate response (non-streaming)
     */
    public String generate(String prompt, int maxTokens, double temperature) {
        if (!isReady()) {
            return "Error: Model not loaded";
        }

        try {
            String config = buildGenerationConfig(maxTokens, temperature);
            return nativeGenerate(engineHandle, prompt, config);
        } catch (Exception e) {
            Log.e(TAG, "Generation error: " + e.getMessage());
            return "Error: " + e.getMessage();
        }
    }

    /**
     * Generate response with streaming
     */
    public void generateStreaming(String prompt, int maxTokens, double temperature, StreamCallback callback) {
        if (!isReady()) {
            callback.onError("Model not loaded");
            return;
        }

        executor.execute(() -> {
            try {
                String config = buildGenerationConfig(maxTokens, temperature);
                nativeGenerateStreaming(engineHandle, prompt, config, new NativeStreamCallback() {
                    StringBuilder fullResponse = new StringBuilder();

                    @Override
                    public void onToken(String token) {
                        fullResponse.append(token);
                        callback.onToken(token);
                    }

                    @Override
                    public void onComplete() {
                        callback.onComplete(fullResponse.toString());
                    }

                    @Override
                    public void onError(String error) {
                        callback.onError(error);
                    }
                });
            } catch (Exception e) {
                callback.onError(e.getMessage());
            }
        });
    }

    /**
     * Build chat messages in OpenAI format
     */
    public String buildChatPrompt(String systemPrompt, String[] history, String userMessage) {
        try {
            JSONObject request = new JSONObject();
            JSONArray messages = new JSONArray();

            // System message
            if (systemPrompt != null && !systemPrompt.isEmpty()) {
                JSONObject sysMsg = new JSONObject();
                sysMsg.put("role", "system");
                sysMsg.put("content", systemPrompt);
                messages.put(sysMsg);
            }

            // History (alternating user/assistant)
            if (history != null) {
                for (int i = 0; i < history.length; i++) {
                    JSONObject msg = new JSONObject();
                    msg.put("role", i % 2 == 0 ? "user" : "assistant");
                    msg.put("content", history[i]);
                    messages.put(msg);
                }
            }

            // New user message
            JSONObject userMsg = new JSONObject();
            userMsg.put("role", "user");
            userMsg.put("content", userMessage);
            messages.put(userMsg);

            request.put("messages", messages);
            return request.toString();

        } catch (JSONException e) {
            Log.e(TAG, "Error building chat prompt: " + e.getMessage());
            return userMessage;
        }
    }

    /**
     * Unload model and free resources
     */
    public void unload() {
        if (engineHandle != 0) {
            nativeDestroyEngine(engineHandle);
            engineHandle = 0;
            isLoaded = false;
            Log.i(TAG, "MLC LLM Engine unloaded");
        }
    }

    /**
     * Get model info
     */
    public String getModelInfo() {
        if (engineHandle != 0) {
            return nativeGetModelInfo(engineHandle);
        }
        return "{}";
    }

    /**
     * Reset conversation/KV cache
     */
    public void resetChat() {
        if (engineHandle != 0) {
            nativeResetChat(engineHandle);
        }
    }

    private String getDefaultConfig() {
        try {
            JSONObject config = new JSONObject();
            config.put("device", "opencl");  // Use GPU via OpenCL
            config.put("model_lib", "");
            return config.toString();
        } catch (JSONException e) {
            return "{}";
        }
    }

    private String buildGenerationConfig(int maxTokens, double temperature) {
        try {
            JSONObject config = new JSONObject();
            config.put("max_tokens", maxTokens);
            config.put("temperature", temperature);
            config.put("top_p", 0.95);
            config.put("frequency_penalty", 0.0);
            config.put("presence_penalty", 0.0);
            config.put("stop", new JSONArray().put("<|endoftext|>").put("<|im_end|>"));
            return config.toString();
        } catch (JSONException e) {
            return "{}";
        }
    }

    public void dispose() {
        unload();
        if (executor != null) {
            executor.shutdown();
        }
    }

    // Native interface callback
    interface NativeStreamCallback {
        void onToken(String token);
        void onComplete();
        void onError(String error);
    }

    // Native methods (implemented in C++)
    private native long nativeCreateEngine(String modelPath, String config);
    private native void nativeDestroyEngine(long handle);
    private native String nativeGenerate(long handle, String prompt, String config);
    private native void nativeGenerateStreaming(long handle, String prompt, String config, NativeStreamCallback callback);
    private native String nativeGetModelInfo(long handle);
    private native void nativeResetChat(long handle);
}

File: LLMClient/Platforms/iOS/MlcLlm/MlcLlmEngine.swift
import Foundation
import Metal
import MetalPerformanceShaders

/// MLC LLM Engine wrapper for iOS
/// Provides high-performance GPU-accelerated LLM inference using Metal
@objc public class MlcLlmEngine: NSObject {

    private var enginePtr: UnsafeMutableRawPointer?
    private var isLoaded: Bool = false
    private var modelPath: String = ""
    private let queue = DispatchQueue(label: "com.llmclient.mlcllm", qos: .userInitiated)

    // Metal device for GPU acceleration
    private var metalDevice: MTLDevice?
    private var commandQueue: MTLCommandQueue?

    // Callbacks
    @objc public var onToken: ((String) -> Void)?
    @objc public var onComplete: ((String) -> Void)?
    @objc public var onError: ((String) -> Void)?
    @objc public var onProgress: ((Double, String) -> Void)?

    @objc public override init() {
        super.init()
        setupMetal()
    }

    private func setupMetal() {
        metalDevice = MTLCreateSystemDefaultDevice()
        if let device = metalDevice {
            commandQueue = device.makeCommandQueue()
            NSLog("[MlcLlmEngine] Metal GPU initialized: \(device.name)")
        } else {
            NSLog("[MlcLlmEngine] Metal not available, will use CPU")
        }
    }

    /// Initialize the engine with a model
    @objc public func initialize(modelId: String, modelPath: String) {
        self.modelPath = modelPath

        queue.async { [weak self] in
            guard let self = self else { return }

            do {
                let config = self.buildConfig()
                self.enginePtr = mlc_create_engine(modelPath, config)

                if self.enginePtr != nil {
                    self.isLoaded = true
                    NSLog("[MlcLlmEngine] Initialized with model: \(modelId)")
                } else {
                    self.onError?("Failed to create MLC engine")
                }
            } catch {
                NSLog("[MlcLlmEngine] Initialization error: \(error.localizedDescription)")
                self.onError?(error.localizedDescription)
            }
        }
    }

    /// Check if engine is ready
    @objc public var isReady: Bool {
        return isLoaded && enginePtr != nil
    }

    /// Generate response (non-streaming)
    @objc public func generate(prompt: String, maxTokens: Int, temperature: Double) -> String {
        guard isReady, let engine = enginePtr else {
            return "Error: Model not loaded"
        }

        let config = buildGenerationConfig(maxTokens: maxTokens, temperature: temperature)

        if let resultPtr = mlc_generate(engine, prompt, config) {
            let result = String(cString: resultPtr)
            mlc_free_string(resultPtr)
            return result
        }

        return "Error: Generation failed"
    }

    /// Generate response with streaming
    @objc public func generateStreaming(prompt: String, maxTokens: Int, temperature: Double) {
        guard isReady, let engine = enginePtr else {
            onError?("Model not loaded")
            return
        }

        queue.async { [weak self] in
            guard let self = self else { return }

            let config = self.buildGenerationConfig(maxTokens: maxTokens, temperature: temperature)
            var fullResponse = ""

            mlc_generate_streaming(engine, prompt, config) { tokenPtr in
                if let ptr = tokenPtr {
                    let token = String(cString: ptr)
                    fullResponse += token

                    DispatchQueue.main.async {
                        self.onToken?(token)
                    }
                }
            }

            DispatchQueue.main.async {
                self.onComplete?(fullResponse)
            }
        }
    }

    /// Build chat prompt with history
    @objc public func buildChatPrompt(systemPrompt: String?, history: [String]?, userMessage: String) -> String {
        var messages: [[String: String]] = []

        // System message
        if let sys = systemPrompt, !sys.isEmpty {
            messages.append(["role": "system", "content": sys])
        }

        // History
        if let hist = history {
            for (index, msg) in hist.enumerated() {
                let role = index % 2 == 0 ? "user" : "assistant"
                messages.append(["role": role, "content": msg])
            }
        }

        // User message
        messages.append(["role": "user", "content": userMessage])

        // Convert to JSON
        do {
            let json: [String: Any] = ["messages": messages]
            let data = try JSONSerialization.data(withJSONObject: json)
            return String(data: data, encoding: .utf8) ?? userMessage
        } catch {
            return userMessage
        }
    }

    /// Unload model
    @objc public func unload() {
        if let engine = enginePtr {
            mlc_destroy_engine(engine)
            enginePtr = nil
            isLoaded = false
            NSLog("[MlcLlmEngine] Engine unloaded")
        }
    }

    /// Reset conversation
    @objc public func resetChat() {
        if let engine = enginePtr {
            mlc_reset_chat(engine)
        }
    }

    /// Get model info
    @objc public func getModelInfo() -> String {
        guard let engine = enginePtr else {
            return "{}"
        }

        if let infoPtr = mlc_get_model_info(engine) {
            let info = String(cString: infoPtr)
            mlc_free_string(infoPtr)
            return info
        }

        return "{}"
    }

    /// Check if Metal GPU is available
    @objc public func isGpuAvailable() -> Bool {
        return metalDevice != nil
    }

    /// Get GPU name
    @objc public func getGpuName() -> String {
        return metalDevice?.name ?? "CPU"
    }

    private func buildConfig() -> String {
        let useGpu = metalDevice != nil
        let config: [String: Any] = [
            "device": useGpu ? "metal" : "cpu",
            "model_lib": ""
        ]

        do {
            let data = try JSONSerialization.data(withJSONObject: config)
            return String(data: data, encoding: .utf8) ?? "{}"
        } catch {
            return "{}"
        }
    }

    private func buildGenerationConfig(maxTokens: Int, temperature: Double) -> String {
        let config: [String: Any] = [
            "max_tokens": maxTokens,
            "temperature": temperature,
            "top_p": 0.95,
            "frequency_penalty": 0.0,
            "presence_penalty": 0.0,
            "stop": ["<|endoftext|>", "<|im_end|>"]
        ]

        do {
            let data = try JSONSerialization.data(withJSONObject: config)
            return String(data: data, encoding: .utf8) ?? "{}"
        } catch {
            return "{}"
        }
    }

    deinit {
        unload()
    }
}

// MARK: - Native C Interface (to be linked with MLC LLM library)
// These are placeholder declarations - actual implementation comes from libmlc_llm.a

@_silgen_name("mlc_create_engine")
func mlc_create_engine(_ modelPath: UnsafePointer<CChar>, _ config: UnsafePointer<CChar>) -> UnsafeMutableRawPointer?

@_silgen_name("mlc_destroy_engine")
func mlc_destroy_engine(_ engine: UnsafeMutableRawPointer)

@_silgen_name("mlc_generate")
func mlc_generate(_ engine: UnsafeMutableRawPointer, _ prompt: UnsafePointer<CChar>, _ config: UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?

@_silgen_name("mlc_generate_streaming")
func mlc_generate_streaming(_ engine: UnsafeMutableRawPointer, _ prompt: UnsafePointer<CChar>, _ config: UnsafePointer<CChar>, _ callback: @escaping (UnsafePointer<CChar>?) -> Void)

@_silgen_name("mlc_reset_chat")
func mlc_reset_chat(_ engine: UnsafeMutableRawPointer)

@_silgen_name("mlc_get_model_info")
func mlc_get_model_info(_ engine: UnsafeMutableRawPointer) -> UnsafeMutablePointer<CChar>?

@_silgen_name("mlc_free_string")
func mlc_free_string(_ str: UnsafeMutablePointer<CChar>)
